### PR TITLE
Gracefully handle product missing from catalog

### LINKED ISSUE #13206: BackorderJob fails on order lines without offer
## Error in Open Food Network (Rails)

**NoMethodError** in **BackorderJob@default**
undefined method `offeredItem' for nil:NilClass

    products = offers.map(&:offeredItem)
                     ^^^^

[View on Bugsnag](https://app.bugsnag.com/yaycode/open-food-network-rails/errors/67d0b13f9aeebb33832cac82?event_id=67d0b13f01131fe9dc410000&i=gh&m=ci)

## Stacktrace

    app/services/fdc_backorderer.rb:94 - map
    app/services/fdc_backorderer.rb:94 - send_order
    app/jobs/backorder_job.rb:125 - place_order
    app/jobs/backorder_job.rb:58 - place_backorder
    app/jobs/backorder_job.rb:27 - block in perform
    app/services/order_locker.rb:18 - block in lock_order_and_variants
    app/services/order_locker.rb:16 - lock_order_and_variants
    app/jobs/backorder_job.rb:26 - perform

[View full stacktrace](https://app.bugsnag.com/yaycode/open-food-network-rails/errors/67d0b13f9aeebb33832cac82?event_id=67d0b13f01131fe9dc410000&i=gh&m=ci)

*Created by Maikel via Bugsnag*

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

- Closes #13206 

#### What should we test?

- Specs should be fine to test this special case.

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: spec/jobs/backorder_job_spec.rb [hunk 1]
```diff
@@ -105,6 +109,27 @@
...
+      )
+
+      stub_request(:get, catalog_url).to_return(body: catalog_json)
+
+      expect {
+        subject.place_backorder(order)
+      }.not_to change { beans.on_hand }
+
+      # And no error was raised, no order was placed.
+    end
```
Comment (rioug) [inline spec/jobs/backorder_job_spec.rb:131]: This only testing a product removed from the catalog, shouldn't we also test the empty backorder and the zero quantity scenario ?

File: app/jobs/backorder_job.rb [hunk 2]
```diff
@@ -55,11 +55,16 @@ def place_backorder(order)
...
+
     place_order(user, order, orderer, backorder)
 
     items.each do |item|
       variant = item.variant
-      variant.on_hand += ordered_quantities[item] if variant.on_demand
+      quantity = ordered_quantities[item]
+      next if quantity.zero?
+
+      variant.on_hand += quantity if variant.on_demand
```
Comment (dacook) [inline app/jobs/backorder_job.rb]: Wouldn't it be just as efficient to let `variant.on_hand += 0` run, rather than adding a check?

**Review by dacook — APPROVED** [review]

File: spec/jobs/backorder_job_spec.rb (hunk 1 above)
Comment (dacook) [reply spec/jobs/backorder_job_spec.rb:131]: I can imagine the shop manager would like to know that the backorder failed.
If they are listing a product for sale, that they can no longer order from supplier, then they should probably remove the product from their shopfront or take some other action.
So would it be worth sending an email to the shop managers here?

File: app/jobs/backorder_job.rb (hunk 2 above)
Comment (mkllnk) [reply app/jobs/backorder_job.rb]: No, it triggers several update statements.

File: spec/jobs/backorder_job_spec.rb (hunk 1 above)
Comment (mkllnk) [reply spec/jobs/backorder_job_spec.rb:131]: Maybe. The latest discussion with Garethe revealed that local stock can still be sold. So a new case to consider in my other PR:

* Product is on-demand and imported to DFC.
* Product is mapped to wholesale variant.
* At the end of one cycle, OFN still has 5 retail items left in stock.
* At the next cycle, the product disappears from the catalog.
* The product needs to switch to stock-controlled and keep the 5 in stock.
* The product can still be sold until sold out.
* We don't need to send an error message in this case. At least it hasn't been requested.
* The item may come back into the catalog, in which case we can resume ordering.

Comment (mkllnk) [reply spec/jobs/backorder_job_spec.rb:131]: > This only testing a product removed from the catalog, shouldn't we also test the empty backorder and the zero quantity scenario ?

Empty backorder: this is tested here as well. No backorder is placed, otherwise there would be an HTTP request to place an order which would fail as we didn't stub anything.

Zero quantity: This is covered as well. When the catalog is missing the product, we return the zero quantity and then we skip that product.

Comment (rioug) [reply spec/jobs/backorder_job_spec.rb:131]: thanks for the clarification !

**Review by rioug — APPROVED** [review]
